Bug: CPU-only build of GMRES solver tries to allocate on GPU

### Description how to reproduce the bug

Error messages are returned when running GMRES tests suggesting the solver is trying to access GPU memory. These are benign messages as computation is carried out to completion on a CPU device and all GPU calls are ignored, but the code should not be trying to access the the device memory in the CPU mode in the first place.

A typical output looks like this:
```
test 15
      Start 15: sys_rand_count_gmres_cgs2_test

15: Test command: /Users/55y/src/resolve2/build/tests/functionality/sys_rand_gmres_test.exe "-x" "no" "-i" "randgmres" "-g" "cgs2" "-s" "count"
15: Working Directory: /Users/55y/src/resolve2/build/tests/functionality
15: Test timeout computed to be: 10000000
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: Results for randomized GMRES solver
15:      Sketching method:               count sketching
15:      Orthogonalization method:       reorthogonalized classical Gram-Schmidt
15:      Hardware backend:               CPU
15:      Solver tolerance:               1e-12
15:      Initial residual norm          ||b-A*x||       : 7.9627092123221482e+03
15:      Initial relative residual norm ||b-A*x||/||b|| : 1.0144607573212390e+00
15:      Final residual norm            ||b-A*x||       : 4.5288034907789973e-09
15:      Final relative residual norm   ||b-A*x||/||b|| : 5.7697616433175111e-13
15:      Number of iterations                           : 19
15: 
15: Test PASSED
15: 
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15: [ERROR] Trying to delete on a GPU device, but GPU support not available.
15/38 Test #15: sys_rand_count_gmres_cgs2_test ........   Passed    0.08 sec
```


### Re::Solve version

`develop` at 36e9dc6ebc8fa862feb4e18b3fae8e17758b8c88

### System and environment details

- MacOS 15.5
- gcc 14.2.0

### Additional information

_No response_

--- a/resolve/vector/Vector.cpp
+++ b/resolve/vector/Vector.cpp
@@ -38,8 +38,8 @@ namespace ReSolve { namespace vector {
    */
   Vector::~Vector()
   {
-    if (owns_cpu_data_) delete [] h_data_;
-    if (owns_gpu_data_) mem_.deleteOnDevice(d_data_);
+    if (owns_cpu_data_ && h_data_) delete [] h_data_;
+    if (owns_gpu_data_ && d_data_) mem_.deleteOnDevice(d_data_);
   }
 
 

--- a/tests/functionality/CMakeLists.txt
+++ b/tests/functionality/CMakeLists.txt
@@ -112,8 +112,8 @@ add_test(NAME rand_gmres_test    COMMAND $<TARGET_FILE:rand_gmres_test.exe>)
 
 if(RESOLVE_USE_KLU)
   # Using KLU as refactorization solver
-  add_test(NAME klu_klu_test    COMMAND $<TARGET_FILE:klu_klu_test.exe> "${test_data_dir}" "-d" "${test_data_dir}")
-  add_test(NAME klu_klu_ir_test COMMAND $<TARGET_FILE:klu_klu_test.exe> "${test_data_dir}" "-d" "${test_data_dir}" "-i")
+  add_test(NAME klu_klu_test    COMMAND $<TARGET_FILE:klu_klu_test.exe> "-d" "${test_data_dir}")
+  add_test(NAME klu_klu_ir_test COMMAND $<TARGET_FILE:klu_klu_test.exe> "-d" "${test_data_dir}" "-i")
 
   # CUDA-SDK specific tests
   if(RESOLVE_USE_CUDA)

--- a/tests/functionality/testKlu.cpp
+++ b/tests/functionality/testKlu.cpp
@@ -95,7 +95,6 @@ int runTest(int argc, char *argv[], std::string& solver_name)
     return -1;
   }
   ReSolve::matrix::Csr* A = ReSolve::io::createCsrFromFile(mat1);
-  A->syncData(ReSolve::memory::DEVICE);
   mat1.close();
 
   // Read first rhs vector
@@ -108,7 +107,6 @@ int runTest(int argc, char *argv[], std::string& solver_name)
   real_type* rhs = ReSolve::io::createArrayFromFile(rhs1_file);
   vector_type vec_rhs(A->getNumRows());
   vec_rhs.copyDataFrom(rhs, ReSolve::memory::HOST, ReSolve::memory::HOST);
-  vec_rhs.syncData(ReSolve::memory::DEVICE);
   rhs1_file.close();
 
   // Allocate the solution vector
@@ -162,7 +160,6 @@ int runTest(int argc, char *argv[], std::string& solver_name)
     return -1;
   }
   ReSolve::io::updateMatrixFromFile(mat2, A);
-  A->syncData(ReSolve::memory::DEVICE);
   mat2.close();
 
   // Load the second rhs vector
@@ -174,7 +171,7 @@ int runTest(int argc, char *argv[], std::string& solver_name)
   }
   ReSolve::io::updateArrayFromFile(rhs2_file, &rhs);
   rhs2_file.close();
-  vec_rhs.copyDataFrom(rhs, ReSolve::memory::HOST, ReSolve::memory::DEVICE);
+  vec_rhs.copyDataFrom(rhs, ReSolve::memory::HOST, ReSolve::memory::HOST);
 
   status = KLU.refactorize();
   error_sum += status;

